jsparagus/actions.py

```diff
@@ -259,18 +320,135 @@ def __init__(self, actions):
         self.actions = tuple(actions)   # Ordered list of actions to execute.
         assert all([not a.is_condition() for a in actions[1:]])
         assert all([not a.update_stack() for a in actions[:-1]])
+
     def __str__(self):
         return "{{ {} }}".format("; ".join(map(str, self.actions)))
+
     def __repr__(self):
         return "Seq({})".format(repr(self.actions))
+
     def is_condition(self):
         return self.actions[0].is_condition()
+
     def condition(self):
         return self.actions[0]
+
     def update_stack(self):
         return self.actions[-1].update_stack()
+
     def reduce_with(self):
         return self.actions[-1].reduce_with()
+
     def shifted_action(self, shift):
         actions = list(map(lambda a: a.shifted_action(shift), self.actions))
         return Seq(actions)
+
+    def contains_accept(self):
+        return any(a.contains_accept() for a in self.actions)
+
+class SeqBuilder:
+    """Aggregate multiple actions in one sequence. Reduce actions added to this
+    sequence are implicitly removed except for the last one but are still
+    considered for rewriting arguments of FunCall. """
+
+    def __init__(self):
+        # offset_stack is used for rewritting offsets which are matching the
+        # value of non-terminals.
+        self.offset_stack = [1] # list(reversed(range(1, 16)))
```

Comment:
The comment for offset_stack should state the invariant. Knowing what it is used for is nice, but I am still in the dark -- `self.offset_stack[i]` gives the index of what, exactly? And what is `i`? I think the explanation needs to use a word like "current" or "so far", since this field changes as we add actions.